### PR TITLE
feat: type declarations in core package

### PR DESCRIPTION
**Pull request by tpluscode**
I'm a little tired of maintaining DTS package for barnard59 so I start with updating the core package to become TS

I kept all test files as plain JS because that greatly reduces the cost to migrate and does not pose any problems

### CODE REVIEW

File: packages/core/lib/Pipeline.ts [hunk 1]
```diff
@@ -115,7 +147,7 @@ class Pipeline extends StreamObject {
 
   // stream interface
 
-  async _read(size) {
+  async _read(size: number) {
     try {
       await this.init()
```
Comment (giacomociti) [inline packages/core/lib/Pipeline.ts:152]: not related to this PR, but my IDE suggests that `await` is not needed

File: packages/core/lib/Pipeline.ts [hunk 2]
```diff
@@ -144,22 +176,27 @@ class Pipeline extends StreamObject {
...
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    } catch (err: any) {
       this._span.recordException(err)
       this.lastChild.stream.destroy(err)
     }
   }
 
-  async _write(chunk, encoding, callback) {
+  async _write(chunk: unknown, encoding: string, callback: (error?: Error | null) => void) {
     await this.init()
```
Comment (giacomociti) [inline packages/core/lib/Pipeline.ts:187]: not related to this PR, but my IDE suggests that `await` is not needed

File: packages/core/lib/Pipeline.ts [hunk 3]
```diff
@@ -144,22 +176,27 @@ class Pipeline extends StreamObject {
...
 
+    assertWritable(this.firstChild)
+
     this._span.addEvent('write')
     return this.firstChild.stream.write(chunk, encoding, callback)
   }
 
-  async _final(callback) {
+  async _final(callback: (error?: Error | null) => void) {
     await this.init()
```
Comment (giacomociti) [inline packages/core/lib/Pipeline.ts:196]: not related to this PR, but my IDE suggests that `await` is not needed

**Review by ludovicm67 — APPROVED** [review]

File: packages/core/lib/Pipeline.ts (hunk 1 above)
Comment (tpluscode) [reply packages/core/lib/Pipeline.ts:152]: Actually, it was related as you explained. Thanks!

**Pull request merged**